test-data/unit/fine-grained.test

```diff
@@ -4854,3 +4854,321 @@ class C:
 ==
 ==
 a.py:4: error: Too many arguments for "C"
+
+[case testOverloadsSimpleFrom]
+import a
+[file a.py]
+import mod
+def f() -> None:
+    x: str = mod.f(str())
+[file mod.py]
+from typing import overload
+@overload
+def f(x: int) -> None: pass
+@overload
+def f(x: str) -> str: pass
+def f(x):
+    pass
+[file mod.py.2]
+from typing import overload
+@overload
+def f(x: int) -> None: pass
+@overload
+def f(x: str) -> int: pass
+def f(x):
+    pass
+[out]
+==
+a.py:3: error: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+[case testOverloadsSimpleTo]
+import a
+[file a.py]
+from typing import overload, Any
+import mod
+def outer() -> None:
+    @overload
+    def f(x: int) -> None: pass
+    @overload
+    def f(x: str) -> str: pass
+    def f(x: Any) -> Any:
+        y: int = mod.f()
+[file mod.py]
+def f() -> int:
+    pass
+[file mod.py.2]
+def f() -> str:
+    pass
+[out]
+==
+a.py:9: error: Incompatible types in assignment (expression has type "str", variable has type "int")
+
+[case testOverloadsRemovedOverload]
+import a
+[file a.py]
+import mod
+def f() -> None:
+    x: str = mod.f(str())
+[file mod.py]
+class C: pass
+from typing import overload
+@overload
+def f(x: int) -> None: pass
+@overload
+def f(x: str) -> str: pass
+@overload
+def f(x: C) -> int: pass
+def f(x):
+    pass
+[file mod.py.2]
+class C: pass
+from typing import overload
+@overload
+def f(x: int) -> None: pass
+@overload
+def f(x: C) -> int: pass
+def f(x):
+    pass
+[out]
+==
+a.py:3: error: No overload variant of "f" matches argument types [builtins.str]
+
+[case testOverloadsDeleted]
+import a
+[file a.py]
+import mod
+def f() -> None:
+    x: str = mod.f(str())
+[file mod.py]
+from typing import overload
+@overload
+def f(x: int) -> None: pass
+@overload
+def f(x: str) -> str: pass
+def f(x):
+    pass
+[file mod.py.2]
+from typing import overload
+[out]
+==
+a.py:3: error: "object" has no attribute "f"
+
+[case testOverloadsUpdatedTypeRecheckImplementation]
+import a
+[file a.py]
+from typing import overload
+import mod
+def outer() -> None:
+    @overload
```

Comment:
I'd rather have a non-nested overload since it's by far the most common scenario.